### ISSUE TITLE
Can't make sure property

### ISSUE THREAD

**Issue by Muhzinmai** [microsoft/TypeScript#45119]
# Bug Report

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAQ1QCggUwDbYFyLgDWYcA7mAJQEBucMAJogN4BQiiMwGO2iAZP0RQAngAdMcYIiy5EAXnmIA5HABGAK0zRlAobOwA6ABbIAzgHkKABQBOcCbdFplDGGbHZkI5ZT0q3Dy8fTiQDPzZ2DhkEMzhsTENsOABzHlxDQM9vSiiAX1Y8oA

Expected : no error

### 🔎 Search Terms

Verify Key in Object

### 🕗 Version & Regression Information

Latest

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAQ1QCggUwDbYFyLgDWYcA7mAJQEBucMAJogN4BQiiMwGO2iAZP0RQAngAdMcYIiy5EAXnmIA5HABGAK0zRlAobOwA6ABbIAzgHkKABQBOcCbdFplDGGbHZkI5ZT0q3Dy8fTiQDPzZ2DhkEMzhsTENsOABzHlxDQM9vSiiAX1Y8oA)

### 💻 Code

```ts

function aaa(cell: unknown): void {
  if(cell && typeof cell == 'object' && cell.hasOwnProperty('display') && 'display' in cell) {

    console.log(cell.display)
  }
}

```

### 🙁 Actual behavior

Property 'display' does not exist on type 'object'.(2339)

### 🙂 Expected behavior

No Error

**Comment by gabrielurbina** [comment]
Adding more details over here, it appears to be a limitation of the [in operator narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) with `object` type. Not sure if this is a defect or a missing feature
```typescript
function aaa(cell: object): void {
  if('display' in cell) {
    console.log(cell.display) // Property 'display' does not exist on type 'object'.(2339)
  }
}
```

**Comment by gabrielurbina** [comment]
as a workaround you can use a typeguard
```typescript
function implementsDisplay(value: unknown): value is { display: unknown} {
  return Boolean(value && typeof value === 'object' && value.hasOwnProperty('display'));
}

function aaa(cell: unknown): void {
  if(implementsDisplay(cell)) {

    console.log(cell.display)
  }
}


```

**Comment by Asaf-S** [comment]
> as a workaround you can use a typeguard
> 
> ```ts
> function implementsDisplay(value: unknown): value is { display: unknown} {
>   return Boolean(value && typeof value === 'object' && value.hasOwnProperty('display'));
> }
> 
> function aaa(cell: unknown): void {
>   if(implementsDisplay(cell)) {
> 
>     console.log(cell.display)
>   }
> }
> ```

Didn't know the `is` operator so far - so thx for that 😊

Your suggestion is a normal [Type Assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions), that can be done in some other ways as well. Some might be more suitable in this case (depending on the context):
```typescript
function aaa(cell: unknown): void {
  if(cell && typeof cell == 'object' && cell.hasOwnProperty('display') && 'display' in cell) {
    console.log((cell as { display: unknown}).display)
  }
}
```

**Comment by gabrielurbina** [comment]
Typescript has a limitation regarding `object.hasOwnProperty`, you can't really tell appart an own property, than a prototype property, so in that aspect, it's stalemate.

On the *in* operator side is where the problem lies

it can't narrow the `object` type 
```typescript
declare  const obj: object;
        if ("test" in obj) { // the in operator is unable to narrow the `object` type
            const j: { test: unknown} = obj; // Property 'test' is missing in type '{}' but required in type '{ test: unknown; }'.
        }
```
it can't narrow the a `Record<string, unknown>` type 
```typescript
declare  const obj: Record<string, unknown>;
        if ("test" in obj) { // the in operator is unable to narrow the Record<string, unknown> type
            const j: { test: unknown} = obj; // Property 'test' is missing in type '{}' but required in type '{ test: unknown; }'.
        }
```
however it does narrow union types 
```typescript
declare  const obj: { test: string} | { best: number};
        if ("test" in obj) { // the in operator is able to narrow the union  { test: string} | { best: number}
            const rec: { test: unknown} = obj; 
        }
```

**Comment by RyanCavanaugh** [comment]
See #21732

**Comment by geoffreytools** [comment]
Could you explain why `hasOwnProperty` is not of the following type?
```typescript
hasOwnProperty<P extends PropertyKey>(this: object, v: P): this is {[K in P]: unknown};
```
The point about `object.hasOwnProperty` being stalemate is unclear. The above would work.
